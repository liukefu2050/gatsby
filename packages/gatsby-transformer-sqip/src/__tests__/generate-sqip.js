import { resolve } from "path"

import { exists, readFile, writeFile } from "fs-extra"
import { sqip } from "sqip"

import generateSqip from "../generate-sqip.js"

jest.mock(`sqip`, () => {
  return {
    sqip: jest.fn(() => {
      return {
        content: Buffer.from(`<svg><!-- Mocked SQIP SVG --></svg>`),
        metadata: {
          dataURI: `data:image/svg+xml,dataURI`,
        },
      }
    }),
  }
})

jest.mock(`fs-extra`, () => {
  return {
    exists: jest.fn(() => false),
    readFile: jest.fn(() =>
      JSON.stringify({
        svg: `<svg><!-- Cached SVG --></svg>`,
        metadata: {
          dataURI: `data:image/svg+xml,dataURI`,
          anyOther: `metadata-value`,
        },
      })
    ),
    writeFile: jest.fn(),
  }
})

afterEach(() => {
  sqip.mockClear()
  exists.mockClear()
  readFile.mockClear()
  writeFile.mockClear()
})

describe(`gatsby-transformer-sqip`, () => {
  const absolutePath = resolve(
    __dirname,
    `images`,
    `this-file-does-not-need-to-exist-for-the-test.jpg`
  )
  const cacheDir = __dirname

  describe(`generateSqip`, () => {
    it(`not cached`, async () => {
      const cache = {
        get: jest.fn(),
        set: jest.fn(),
      }
      const numberOfPrimitives = 5
      const blur = 0
      const mode = 3
      const result = await generateSqip({
        cache,
        cacheDir,
        absolutePath,
        numberOfPrimitives,
        blur,
        mode,
      })
      expect(result).toMatchSnapshot()

      expect(sqip).toHaveBeenCalledTimes(1)
      const sqipArgs = sqip.mock.calls[0][0]
      expect(sqipArgs.input).toMatch(absolutePath)
      delete sqipArgs.input
      expect(sqipArgs).toMatchSnapshot()

      expect(exists).toHaveBeenCalledTimes(1)
      expect(writeFile).toHaveBeenCalledTimes(1)
      expect(readFile).toHaveBeenCalledTimes(0)
    })
    it(`cached`, async () => {
      exists.mockImplementationOnce(() => true)
      const cache = {
        get: jest.fn(),
        set: jest.fn(),
      }
      const numberOfPrimitives = 5
      const blur = 0
      const mode = 3
      const result = await generateSqip({
        cache,
        cacheDir,
        absolutePath,
        numberOfPrimitives,
        blur,
        mode,
      })

      expect(result).toMatchSnapshot()

      expect(sqip).toHaveBeenCalledTimes(0)

      expect(exists).toHaveBeenCalledTimes(1)
      expect(writeFile).toHaveBeenCalledTimes(0)
      expect(readFile).toHaveBeenCalledTimes(1)
    })
    it(`returns null for unsupported files`, async () => {
      exists.mockImplementationOnce(() => true)

      const cache = {
        get: jest.fn(),
        set: jest.fn(),
      }
      const numberOfPrimitives = 5
      const blur = 0
      const mode = 3
      const result = await generateSqip({
        cache,
        cacheDir,
        absolutePath: absolutePath.replace(`.jpg`, `.svg`),
        numberOfPrimitives,
        blur,
        mode,
      })

      expect(result).toBe(null)

      expect(sqip).toHaveBeenCalledTimes(0)
      expect(exists).toHaveBeenCalledTimes(0)
      expect(writeFile).toHaveBeenCalledTimes(0)
      expect(readFile).toHaveBeenCalledTimes(0)
    })
  })
})
