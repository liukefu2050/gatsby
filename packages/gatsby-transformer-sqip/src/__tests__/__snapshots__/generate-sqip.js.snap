// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gatsby-transformer-sqip generateSqip cached 1`] = `
Object {
  "metadata": Object {
    "anyOther": "metadata-value",
    "dataURI": "data:image/svg+xml,dataURI",
  },
  "svg": "<svg><!-- Cached SVG --></svg>",
}
`;

exports[`gatsby-transformer-sqip generateSqip not cached 1`] = `
Object {
  "dataURI": "data:image/svg+xml,dataURI",
  "metadata": Object {},
  "svg": "<svg><!-- Mocked SQIP SVG --></svg>",
}
`;

exports[`gatsby-transformer-sqip generateSqip not cached 2`] = `
Object {
  "plugins": Array [
    Object {
      "name": "sqip-plugin-primitive",
      "options": Object {
        "mode": 3,
        "numberOfPrimitives": 5,
      },
    },
    "sqip-plugin-data-uri",
    "sqip-plugin-svgo",
  ],
}
`;
